### PR TITLE
OOP exercise

### PR DESCRIPTION
**Pull request by BaoBao-provjpno1**

### CODE REVIEW

File: abstraction.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    public area(): number{
+        return this.side*4;
+    }
+    public perimeter(): number{
+        return this.side*this.side;
+    }
+
+}
+class Circle extends Shape{
```
Comment (kahnext) [inline abstraction.ts]: split class into specific one

File: abstraction.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+abstract class Shape{
+    abstract area(): number;
+    abstract perimeter(): number;
+}
+class Square extends Shape{
+    private side : number;
+    
+    constructor(side : number){
+        super()
```
Comment (kahnext) [inline abstraction.ts]: ending character missing

File: abstraction.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        this.side = side;
+    }
+
+    public area(): number{
+        return this.side*4;
+    }
+    public perimeter(): number{
+        return this.side*this.side;
```
Comment (kahnext) [inline abstraction.ts]: code format

File: polymorAndinhen.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+}
+class Cat extends Animal{
+    public speak(): void{
+     console.log(`the ${this.name} say meow meow meow meow meow meow meow meow meow`);
+    }
+}
+
+
+//đa hình
```
Comment (kahnext) [inline polymorAndinhen.ts]: wrong polymorphism concept

File: poly_n_andi/animal.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+//animal.ts
+class Animal {
+  protected name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+  //speak
+  public speak(): void {
```
Comment (kahnext) [inline poly_n_andi/animal.ts]: don't see overloading support

Comment (BaoBao-provjpno1) [reply poly_n_andi/animal.ts]: thank you , the code was updated.

File: encapsulation/bank.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      this.balance -= amount;
+      console.log(`withdrew: ${amount}, new balance: ${this.balance}`);
+    } else console.log(`Insufficient funds`);
+  }
+
+  public getBalance(): number {
+    return this.balance;
+  }
+}
+const acount = new bankAccount("YoungBooprovjp", 9000);
```
Comment (kahnext) [inline encapsulation/bank.ts]: do not call Unit test inside, make main.ts entry instead

File: abstraction/src/square.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  }
+  //perimeter
+  public perimeter(): number {
+    return this.side * this.side;
+  }
+  //area overloading
+  public area(): number;
+  public area(a:string): string;
+
+  public area(a?:string): string | number{
```
Comment (kahnext) [inline abstraction/src/square.ts]: good overloading support

File: encapsulation/bank.ts (hunk 6 above)
Comment (BaoBao-provjpno1) [reply encapsulation/bank.ts]: thank you, I have updated the code.

File: abstraction.ts (hunk 1 above)
Comment (BaoBao-provjpno1) [reply abstraction.ts]: fixed

File: abstraction.ts (hunk 2 above)
Comment (BaoBao-provjpno1) [reply abstraction.ts]: fixed

File: polymorAndinhen.ts (hunk 4 above)
Comment (BaoBao-provjpno1) [reply polymorAndinhen.ts]: fixed

File: encapsulation/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+## Getting Started
```
Comment (kahnext) [inline encapsulation/README.md]: keep only readme file for whole project

File: abstraction.ts (hunk 3 above)
Comment (BaoBao-provjpno1) [reply abstraction.ts]: fixed

File: encapsulation/README.md (hunk 8 above)
Comment (BaoBao-provjpno1) [reply encapsulation/README.md]: I fixed, please check new version

**Review by kahnext — APPROVED** [review]
LGTM

**Pull request merged**